Fix #3836: Upgrade training interface to show unresolved answers.

prasanna08 opened the pull request:
This PR implements 2nd milestone of the ML training interface project. It implements following things:
* A new teach Oppia modal to show unresolved answers.
* A new training data editor which allows modifications to the training data and adding new answers to training data (fixing #3170).

Few screenshots of new interface:
## Teach Oppia Modal

>  ![teach oppia modal](https://user-images.githubusercontent.com/8678373/41325990-afbb55ca-6eda-11e8-9d77-bb11cd62a09a.png)


### Upon clicking `Fix` button

> ![teach oppia modal 2](https://user-images.githubusercontent.com/8678373/41326067-e8b07d1a-6eda-11e8-9517-a8fc3b8ca140.png)


### When all of the unresolved answers have been addressed.

> ![teach oppia modal 1](https://user-images.githubusercontent.com/8678373/41326079-f5128d78-6eda-11e8-87e8-7e03890cd4ac.png)


## Training Data Editor

> ![training data editor 1](https://user-images.githubusercontent.com/8678373/41326084-01825b9c-6edb-11e8-8941-d4f6035814eb.png)


### Upon clicking on any answer of training data

> ![training data editor 4](https://user-images.githubusercontent.com/8678373/41326096-0dfb059a-6edb-11e8-8f18-4994fb88015b.png)


### `Add New Answer` tab

> ![training data editor 2](https://user-images.githubusercontent.com/8678373/41326104-1aa5c8a2-6edb-11e8-8608-2dbf0c931e70.png)

### When new added answer is explicitly classified (using either hard rules or training data classification)

> ![training data editor 3](https://user-images.githubusercontent.com/8678373/41326124-33da018a-6edb-11e8-8543-7406f9fef98d.png)

## Answer Group Editor
It now provides `Modify Training Data` button.

> ![new answer group editor](https://user-images.githubusercontent.com/8678373/41326151-4f7726d4-6edb-11e8-9245-a35cc50b66f4.png)


-------------------------------------------------------------------------------------------------------------------------------------------
### Enable interface
1. To enble the new training interface flip following constant in feconf.py file.
```
  SHOW_TRAINABLE_UNRESOLVED_ANSWERS
```
2.Create an exploration with TextInput / CodeRepl interaction. It should work only for them as they are only trainable interactions available in Oppia.

prasanna08 committed:
Interface work in progress

prasanna08 committed:
Finish up training interface

prasanna08 committed:
Merge branch 'develop' of github.com:oppia/oppia into training_interface

prasanna08 committed:
Final training interface

prasanna08 committed:
Fix Liniting issues

prasanna08 committed:
Fix frontend tests

prasanna08 committed:
Use material icons instead of glyphicons

prasanna08 committed:
Fix comments

prasanna08 commented:
Hi, @seanlip implemented 2nd milestone of training interface [project](https://docs.google.com/document/d/19v-zTFS7_8nysUAggDau3xCDN_WS4x20heJJjpZdALs/edit).

prasanna08 committed:
Fix backend and frontend test

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateInteraction.js:
```diff
@@ -440,6 +442,10 @@ oppia.directive('testInteractionPanel', [
             });
           };
+
+          $scope.submitAnswerFromButton = function() {
```
Probably better to rename to onSubmitAnswerFromButton.

Just to check, though, why was this changed? (On the surface it seems unrelated to this PR; was it an existing bug?)

prasanna08 replied:
Yep, it's an existing bug. There was no submit button. So pressing `enter` was `submit` but that does not work for complex interactions like code repl where `enter` is a newline. So I added a submit button.

prasanna08 replied:
Actually, you would find a few more things changed / removed / added. The original thing was all broken up.

seanlip replied:
Ah, got it, thanks -- I think I missed that this was in testInteractionPanel, I thought it was something showing up in the UI already.

seanlip reviewed (commented):
Reviewed until https://github.com/oppia/oppia/pull/5098/files#diff-c16e3cb3333a1b0eff1241c171d72084L280, just one question so far. Will continue tomorrow!

prasanna08 committed:
Addressed review comments

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
+              }).success(function(response) {
+                $scope.showUnresolvedAnswers(response.unresolved_answers);
+              }).error(function(response) {
+                $scope.showUnresolvedAnswers([]);
```
Shouldn't some kind of error be displayed here (or at least a $log.error() to record this in the backend)?

prasanna08 replied:
Done. Added $log.error()

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
+                  classificationResult.outcome.feedback.getHtml());
+                var classificationType = (
+                  classificationResult.classificationCategorization);
+                if (classificationType !== EXPLICIT_CLASSIFICATION &&
```
Consider moving answerTemplate and feedbackHtml into this 'if' brace, since you won't need them otherwise.

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
+              });
+            };
+
+            $scope.confirmToTrainingData = function(answerIndex) {
```
I'm not sure what this function does based on its name. Could you elaborate? What does "confirm to" mean?

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
+              $scope.unresolvedAnswers.splice(answerIndex, 1);
+              if (answer.classificationResult.classificationCategorization ===
+                  DEFAULT_OUTCOME_CLASSIFICATION) {
+                TrainingDataService.trainDefaultResponse(answer.answer);
```
It's not clear to me what trainDefaultResponse() means -- does it mean add the answer to the "default response" answer group? If so, maybe associateWIthDefaultResponse / associateWithAnswerGroup might be clearer?

(Also, normally I think we talk more about "training an ML model" rather than "training an answer group".)

prasanna08 replied:
Yep. `trainDefaultResponse` adds the answer to confirmed unclassified answers. I'll modify the names.

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
+                answer.classificationResult.answerGroupIndex, answer.answer);
+            };
+
+            $scope.finishTrainingAnswer = function() {
```
It seems strange to represent this a separate step. Consider keeping it as an anonymous function instead (on line 349)?

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
+              $scope.selectedAnswerIndex = answerIndex;
+              answer = $scope.unresolvedAnswers[answerIndex].answer;
+              return TrainingModalService.openTrainUnresolvedAnswerModal(
+                answer, true, $scope.finishTrainingAnswer);
```
Is it possible to dismiss the modal without training the answer? If so what happens to $scope.selectedAnswerIndex?

prasanna08 replied:
It stays as it is until a new answer is selected by the creator. An answer will only be removed when it is successfully assigned by the creator.

seanlip replied:
But if the modal is dismissed shouldn't we reset $scope.selectedAnswerIndex to null (since no answer would be selected at that time)?

prasanna08 replied:
But that would require adding another callback. It does no harm even if we keep it as it is.

seanlip replied:
Ah -- I was going to object to that, because it's conceptually muddy. If no answer is selected then we shouldn't have $scope.selectedAnswerIndex indicating otherwise.

However, I just looked at this more carefully, and I don't see the point of this being declared on $scope in the first place. If you put $scope.finishTrainingAnswer inline, then I think you can just have `var selectedAnswerIndex` and then that variable just becomes internal to this function. That seems like it would be a lot clearer.

prasanna08 replied:
Ok, will do.

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateResponses.js:
```diff
@@ -259,99 +250,113 @@ oppia.controller('StateResponses', [
-                  classificationResult.answerGroupIndex);
-              }
-            };
+            $scope.showLoadingDots = true;
```
$scope.loadingDotsAreShown (to avoid starting with a verb when the variable does not represent a function)

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+
+/**
+ * @fileoverview Service which handles opening and closing
+ * the training data of an answer group.
```
This is unclear -- do you mean opening and closing the training data editor for an answer group? (Not sure what it means to open and close "training data".)

prasanna08 replied:
Yep, training data editor of an answer group.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+  '$rootScope', '$uibModal', 'AlertsService', 'UrlInterpolationService',
+  function($rootScope, $uibModal, AlertsService, UrlInterpolationService) {
+    return {
+      openTrainingDataEditor: function(externalSave) {
```
Add docstring/jsdoc explaining what externalSave means.

prasanna08 replied:
It saves the modified state content.

seanlip replied:
Yep -- add JSDoc specifying its typeinfo, description, etc. (so that devs know exactly what it means to call this function with that arg set to true vs false). It's not clear from the name.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                ExplorationHtmlFormatterService.getInteractionHtml(
+                  stateInteractionIdService.savedMemento,
+                  stateCustomizationArgsService.savedMemento,
+                  false, 'testInteractionInput'));
```
Worth pulling out 'testInteractionInput' into a named constant?

prasanna08 replied:
It's for focus target and so basically used for focusing on tag named 'test-interaction-input'. I don't think it should be in a constant since it is a tag name.

seanlip replied:
Are you sure it is meant to be a tag name and not an arbitrary string? This is different from how I remember focusService working, I thought it would focus based on the value of a customizable attribute.

Otherwise we would be assuming that a tag _type_ is unique in the DOM for a given page, which is a pretty strong assumption.

prasanna08 replied:
This is what ExplorationHtmlFormatterService does
```
element.attr('label-for-focus-target', labelForFocusTarget);
```

where `labelForFocusTarget` is the last argument and that string is indeed arbitrary string. Basically used to guide FocusManagerService. So you are correct. I will put it in a constant and add FocusManagerService (seems like I forgot to add that, too).

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                $scope.trainingData = [];
+                TrainingDataService.getTrainingDataOfAnswerGroup(
+                  answerGroupIndex).forEach(function(answer) {
+                  answerTemplate = (
```
missing var?

prasanna08 replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                      answer, stateInteractionIdService.savedMemento,
+                      stateCustomizationArgsService.savedMemento));
+                  $scope.trainingData.push({
+                    answer: answer,
```
can we make a domain object for this?

prasanna08 replied:
It is created only temporarily and also not used outside of this service. I don't think its worth creating a domain object since it is nothing concrete, just a temporarily created data structure so that view can be generated.

seanlip replied:
Hm, technically, we're now planning to create domain objects for anything that's represented as a dict with fixed keys, just as a matter of principle. The problem with passing around raw dicts is that they tend to get modified and you then never really know what's in them at any given stage in time. (And it always starts off being a local small thing with limited use -- until it isn't.)

So -- still leaning towards a domain object, but I do take your point. Consider the above information as an additional data point, and feel free to make the call.

prasanna08 replied:
I don't think that having a domain object is worth it since this structure is created only when the specific modal is opened (it's not even on the page). So I am skipping this.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                  answerTemplate = (
+                    ExplorationHtmlFormatterService.getAnswerHtml(
+                      answer, stateInteractionIdService.savedMemento,
+                      stateCustomizationArgsService.savedMemento));
```
Just to check, what happens to the training data if the customization args change, say (e.g. if we delete/rearrange choices in MultipleChoiceInput)?

I guess that might not be an issue because we don't need ML for MultipleChoiceInput, but I couldn't find the place where this PR depends on feconf.INTERACTION_CLASSIFIER_MAPPING.

prasanna08 replied:
That case has not been handled in the current infrastructure. Also this PR does not depend on `feconf.INTERACTION_CLASSIFIER_MAPPING` because that mapping used when we train a classifier on training data. But training data will there only when interaction is trainable. So if some interaction is not trainable then no part of this interface will be enabled. So no training data modification will be allowed in that case.

seanlip replied:
Ah, OK, that's interesting.

Can we make the presence of all this stuff depend on feconf.INTERACTION_CLASSIFIER_MAPPING (in addition to whether there are answers to train)? I'm thinking about the case where we expose a trainable interaction that we later decide is non-trainable, and it seems like a good idea to me to have the additional check as a safeguard (defense-in-depth).

prasanna08 replied:
The interaction classifier mapping is separate from training interface. Its purpose is to provide info whether a classifier on Oppia-ml is available to train the training data of given interaction. I think that case is entirely separate from whether interaction should have training data. Also, the mapping is used between Oppia and Oppia-ml and is not transferred to the frontend. I think the check for isInteractionTrainable() is sufficient.

Additionally, it keeps things loosely coupled. We can always make an interaction non-trainable even if we have a classifier on Oppia-ml (like if the classifier is bad or something). Why do you suggest we should have deeper checks?

seanlip replied:
Perhaps because I missed the check for isInteractionTrainable() -- sorry! Using that would also be fine IMO. Just some way to ensure that this isn't shown for all interactions.

That said, I can't seem to find isInteractionTrainable() when doing Ctrl+F on this page, except for code which has been deleted. Could you please point me to its use?

prasanna08 replied:
The name is `isCurrentInteractionTrainable`. My mistake. Also, take a look at this [line](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/pages/exploration_editor/editor_tab/state_editor_responses.html#L98) in state_editor_responses.html.

So, `Teach Oppia` button and `Modify Training Data` both are guarded using `isCurrentInteractionTrainable()`.

seanlip replied:
Excellent, thanks! Sounds good then.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+              };
+
+              $scope.removeAnswerFromTrainingData = function(answerIndex) {
+                answer = $scope.trainingData[answerIndex].answer;
```
Missing var?

prasanna08 replied:
Fixed.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                  AnswerClassificationService.getMatchingClassificationResult(
+                    _explorationId, _stateName, _state,
+                    newAnswer, rulesService));
+                var newAnswerFeedbackHtml = 'Nothing';
```
This seems weird. Shouldn't it be null or the empty string, and the UI displays it as "Nothing" or "[No feedback specified]" if necessary?

prasanna08 replied:
I don't know. This is how it was in previous interface and so I kept this part as it is.

seanlip replied:
Ah ok. Let's change it to what I suggested? Semantically that would be better, I think, rather than using a sentinel value of 'Nothing'.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                    classificationResult.outcome.feedback.getHtml());
+                }
+                if (newAnswerOutcomeDest === _stateName) {
+                  dest = '<em>(try again)</em>';
```
I don't see where `dest` is declared or used...

Also we shouldn't be constructing HTML in the JS code. See if you can do a UI transform on this instead, similar to my comment above for newAnswerFeedbackHtml.

prasanna08 replied:
Actually, I forgot to replace `dest` with `newAnswerOutcomeDest`. So consider `dest` as `newAnswerOutcomeDest`. Again, this is how it was before so I kept it as it is.

prasanna08 replied:
Fixed. Also removed `em` tags. I think UI looks just fine without it, too.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                // classification results to the creator.
+                if (classificationType === EXPLICIT_CLASSIFICATION ||
+                    classificationType === TRAINING_DATA_CLASSIFICATION) {
+                  $scope.newAnswerIsResolved = true;
```
Maybe $scope.newAnswerIsAlreadyResolved (to indicate that this is already the case, rather than something that the creator has just done)?

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -0,0 +1,165 @@
+                  answer, true, $scope.finishTrainingAnswer);
+              };
+
+              $scope.finishTrainingAnswer = function() {
```
Perhaps get rid of this and just have _rebuildTrainingData inline in line 152? Seems a bit odd to separate it out like this, IMO.

prasanna08 replied:
I thought that this would be faster since we know which answer is being addressed. _rebuild would have to go through entire training data again.

seanlip replied:
Not sure I understand, sorry. Line 156 just calls _rebuildTrainingData() directly. So how is this different then doing something like the following?
```
                return TrainingModalService.openTrainUnresolvedAnswerModal(
                  answer, true, _rebuildTrainingData);
```

prasanna08 replied:
What I meant is that, since we know which answer is being addressed we can remove it from `$scope.trainingData` which would be cheaper than generating `$scope.trainingData` from scratch.

seanlip replied:
I don't have an objection to that, but that doesn't seem to be what $scope.finishTrainingAnswer() is currently doing at the moment either...

(The original change I proposed was meant to be a no-op in terms of functionality -- I was just talking about moving line 156 to line 152.)

prasanna08 replied:
Oops, my bad. I thought we were talking about `finishTrainingAnswer()` method of teach_oppia_modal (the one you suggested to make anonymous function). This is a different case and your original comment stands right. Thanks.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataService.js:
```diff
@@ -82,28 +79,23 @@ oppia.factory('TrainingDataService', [
+        var trainingDataAnswers = [];
         var answerGroups = ResponsesService.getAnswerGroups();
+
+        // Remove the answer from all answer groups.
```
This doesn't seem to be removing anything?

prasanna08 replied:
My bad. Will remove it.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataServiceSpec.js:
```diff
@@ -217,16 +217,6 @@ describe('TrainingDataService', function() {
     ]);
   });
 
-  it('should remove unresolved answers after training', function() {
```
Should add Karma tests for new public methods, here and in other services.

prasanna08 replied:
Added new tests.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingModalService.js:
```diff
@@ -14,15 +14,15 @@
  * @fileoverview Service which handles opening and closing
- * the training modal used for both unresolved answers
- * and answers within the training data of a classifier.
+ * the training modal used for both unresolved answers.
```
strike "both"

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingModalService.js:
```diff
@@ -54,6 +54,11 @@ oppia.factory('TrainingModalService', [
+              };
+
+              $scope.exitTrainer = function() {
+                $uibModalInstance.close();
```
Should this be dismiss()?

prasanna08 replied:
Changed.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/teach_oppia_modal_directive.html:
```diff
@@ -1,55 +1,66 @@
-  <h3>Enter a Sample Answer</h3>
-</div>
+<style>
+.unresolved-answers-loading-text {
```
Can we prefix all of these with the directive name, to make sure that these CSS classes are always scoped only to this modal? (maybe teach-oppia-modal if that's the directive name, or add a .teach-oppia-modal class to line 34)

prasanna08 replied:
Actually, this is not a directive (its modal that JS generates upon clicking button). I will add `teach-oppia-modal` class on line 34.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/teach_oppia_modal_directive.html:
```diff
@@ -1,55 +1,66 @@
+  <div class="unresolved-answers-list oppia-rule-details-header">
+    <p class='unresolved-answers-loading-text' ng-show="showLoadingDots">Loading <loading-dots></loading-dots></p>
+    <div ng-show="unresolvedAnswers.length > 0" style="padding-top: 10px; border-top: 1px solid #e5e5e5;">
+      <div class='unresolved-answer' ng-repeat="answer in unresolvedAnswers track by $index">
```
Here and elsewhere, use double quotes for HTML attrs

/cc @apb7 @kevinlee12

apb7 replied:
Filed Issue #5108 for the same. Thanks!

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/teach_oppia_modal_directive.html:
```diff
@@ -1,55 +1,66 @@
-        <div>
-          <span><strong>This is due to a specific rule and therefore cannot be trained. If you do not think this is right, you should change the rule directly referring to this answer.</strong></span>
+        <div class='unresolved-answer-btn'>
+          <button type='button' class='btn btn-success' ng-click="confirmToTrainingData($index)">
```
Ah. Maybe call it confirmAnswerAssignment() or similar?

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/teach_oppia_modal_directive.html:
```diff
@@ -1,55 +1,66 @@
-  <div>
+    <div ng-show="unresolvedAnswers.length === 0">
+      <span>Congratulations! You've finished addressing important unresolved answers.</span>
+      <br/>
```
As a general rule, try to avoid br-tags to achieve visual effects. Use p-tags and styling (e.g. margins) instead.

prasanna08 replied:
Using `<p>` tags now.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/test_interaction_modal_directive.html:
```diff
@@ -1,14 +1,14 @@
+<style>
+.submit-answer-btn {
```
Indent by 2.

/cc @apb7 @kevinlee12

apb7 replied:
The indent rule has been already added in Issue #5097. I will try to make a fix for this during the weekend since it seems to be occurring at a number of places at the moment.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/test_interaction_modal_directive.html:
```diff
@@ -1,14 +1,14 @@
+  <div>
+    <h3> Enter a sample answer </h3>
+  </div>
+  <br>
```
Ditto re br, here and elsewhere.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/training_data_editor_directive.html:
```diff
@@ -0,0 +1,83 @@
+<style>
+.training-data-answer {
```
Indent all this by 2 (up to line 12).

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/training_data_editor_directive.html:
```diff
@@ -0,0 +1,83 @@
+
+<div class="modal-body">
+  <ul class="nav nav-tabs">
+    <li class='active'><a data-toggle="tab" href="#trainingData">Training Data</a></li>
```
use double quotes for all attrs, here and below

prasanna08 replied:
Done.

seanlip reviewed (changes requested):
Hi @prasanna08, took a full pass at the code (though I haven't looked at the UI yet, I'll do that on the next pass). Thanks!

prasanna08 committed:
Addressed review comments.

prasanna08 committed:
Fix var

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -39,27 +45,28 @@ oppia.factory('TrainingDataEditorPanelService', [
               var answerGroupIndex = (
                 ResponsesService.getActiveAnswerGroupIndex());
 
+              var TEST_INTERACTION_INPUT = 'testInteractionInput';
```
Better to call this FOCUS_LABEL_... (start with the type). Otherwise, without context, hard to tell what this represents conceptually.

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -137,23 +145,19 @@ oppia.factory('TrainingDataEditorPanelService', [
-
-              $scope.finishTrainingAnswer = function() {
-                _rebuildTrainingData();
+                  answer, true, _rebuildTrainingData());
```
Did you test this (at least manually)?

prasanna08 replied:
Ah, yes. I did but I missed this scenario. Thanks for notifying.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingModalService.js:
```diff
@@ -14,13 +14,21 @@
     return {
+      /**
+      * Opens unresolved answer trainer modal for given answer.
+      * @param {Object} - unhandledAnswer The answer to be trained.
```
The dash is in the wrong place.

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingModalService.js:
```diff
@@ -14,13 +14,21 @@
+      * Opens unresolved answer trainer modal for given answer.
+      * @param {Object} - unhandledAnswer The answer to be trained.
+      * @param {Boolean} externalSave - Whether to save the modified training
+          data externally in state.
```
Let's add a bit more context here. Why wouldn't we want to do this? (i.e. why wouldn't it be true)?

prasanna08 replied:
I don't know. As long as I see, it would always be true but it was present in earlier directives and so I thought maybe it has some significance. Should I remove this argument?

seanlip replied:
Yes, if we don't understand why we're doing it :) (And ditto elsewhere.)

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/test_interaction_modal_directive.html:
```diff
@@ -1,14 +1,19 @@
+<div class="form-inline test-interaction-modal" style="margin-bottom: 20px;">
   <div>
-    <h3> Enter a sample answer </h3>
+    <p class='enter-answer-label'> Enter a sample answer </p>
```
Double quotes please. /cc @apb7 @kevinlee12

apb7 replied:
Added in Issue #5108. Thanks!

prasanna08 replied:
Done. Sorry, I am habituated to single quotes.

seanlip reviewed (commented):
Thanks @prasanna08! Did a quick pass, will probably take a more detailed look (including poking at UI) tomorrow.

seanlip commented on core/templates/dev/head/components/answer_group_editor_directive.html:
```diff
@@ -81,6 +81,13 @@
     </button>
+
+    <div ng-if="isCurrentInteractionTrainable()">
+      <br>
```
Try not to use additional br tags for styling. Use margins etc. instead on existing elements.

br tags should semantically indicate a line break; this isn't a line break. It's important that HTML be semantically valid for, e.g., screenreader users.

prasanna08 replied:
Done

seanlip commented:
So, I loaded up the Welcome exploration and submitted a bunch of random stuff to the second state. But when clicking on the "+ Teach Oppia" button, I get "Congratulations! You've finished addressing important unresolved answers. Please visit this page later for more unresolved answers." How do I enable unresolved answers?

seanlip commented:
Also, with regards to the UI, I recommend the following changes:
- Remove the "+" from Teach Oppia; it's not exactly an "add new list item" action.
- Move the Teach Oppia button to the bottom of "Unresolved Student Answers" on the right hand side of the page; it belongs more logically there.

prasanna08 commented:
> So, I loaded up the Welcome exploration and submitted a bunch of random stuff to the second state. But when clicking on the "+ Teach Oppia" button, I get "Congratulations! You've finished addressing important unresolved answers. Please visit this page later for more unresolved answers." How do I enable unresolved answers?

* Make sure that `feconf.SHOW_TRAINABLE_UNRESOLVED_ANSWERS = True`
* Also set `feconf.STATE_ANSWER_STATS_MIN_FREQUENCY = 1`

> Remove the "+" from Teach Oppia; it's not exactly an "add new list item" action.

Done.

> Move the Teach Oppia button to the bottom of "Unresolved Student Answers" on the right hand side of the page; it belongs more logically there.

Done

prasanna08 committed:
Address review comments and UI changes

prasanna08 commented:
Pushed new commit which resolves all comments and following changes have been made in UI:
* Teach Oppia Modal
  * Changed header to "Confirm Oppia's Response"
  * Moved "confirm" and "fix" button to "Response" header.
  * Enabled backdrop (gray background and clicking outside closes modal).
  * Removed arrow from "Response" header.
  * Removed extra line at the begining of modal body.

* Train Unresolved Answer (feedback selection modal)
  * Changed header to "Select Oppia's response"
  * Added confirm button. Now upon selecting answer group creator will be shown "confirm" button rather than directly adding answer to answer group's training data.
  * Now feedback is also shown along with destination when selecting appropriate feedback from list.
  * New response is added to exploration only if that response is selected by creator.

* Training Data Editor Modal
  * Creator is not allowed to remove / reassign a training data answer if there is only one answer (it conflicts when there is only one training data answer and no rule in answer group. So removing the only available training data answer is illegal change and requires removal of answer group).

* Moved teach oppia button under "Issue Overview" side bar.

 I'll update screenshots when I get some free time.

seanlip commented on core/templates/dev/head/pages/exploration_player/AnswerClassificationService.js:
```diff
@@ -111,8 +111,7 @@ oppia.factory('AnswerClassificationService', [
           oldState.interaction.id].is_trainable;
 
-        if (ruleBasedOutcomeIsDefault && interactionIsTrainable &&
-            ENABLE_ML_CLASSIFIERS) {
```
Just to check -- this is supposed to be removed?

prasanna08 replied:
Oops, my mistake. I wanted to make sure that training data classification is allowed irrespective of ml classifiers are enabled or not.

seanlip replied:
Ah, sorry -- so, just to confirm, is the intention to remove it or not?

I think we should have a flag to hide the training data stuff though, at least until the ML stuff is actually working. Otherwise the training data UI components will clutter the UI in a way that's somewhat opaque to creators.

prasanna08 replied:
It will be removed from this line and added to a later line (where the classifier is used). Training data stuff can be hidden from UI by setting `SHOW_TRAINABLE_UNRESOLVED_ANSWERS = False`.

seanlip replied:
OK, got it. Thanks!

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateEditor.js:
```diff
@@ -154,35 +167,40 @@ oppia.directive('trainingPanel', [
-              var index = ResponsesService.getAnswerGroupCount() - 1;
-              TrainingDataService.associateWithAnswerGroup(
-                index, $scope.answer);
+              // Push the new outcome at the end of the all outcomes.
```
all --> existing

prasanna08 replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateEditor.js:
```diff
@@ -141,6 +143,17 @@ oppia.directive('trainingPanel', [
           _updateAnswerTemplate();
+          $scope.selectedAnswerGroupIndex = -1;
+
+          var _saveNewAnswerGroup = function(newAnswerGroup) {
```
@DubeySandeep -- could you please review this part (the content ID)?

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateEditor.js:
```diff
@@ -118,13 +118,15 @@ oppia.directive('trainingPanel', [
             OutcomeObjectFactory, GenerateContentIdService,
-            COMPONENT_NAME_FEEDBACK) {
+            COMPONENT_NAME_FEEDBACK,
+            stateContentIdsToAudioTranslationsService) {
```
Line breaks here should match up with the string versions above, so that it's easy to compare. (We've had issues before with missing dependencies.)

prasanna08 replied:
Done.

seanlip reviewed (changes requested):
Thanks, @prasanna08!

I took it for another spin; some more UI comments:
- Change "Teach Oppia" to "Train Oppia"; the creator is already teaching the student, and this mechanism of training is a bit different from regular teaching.
-  Adjust the buttons so that they lie fully above the first line of the response. Give enough space for the 'Response' header to "breathe" (e.g. add a bit of margin above it), so that neither the answer text or the response text gets cut off by the Confirm/Fix buttons, regardless of how long either the answer or the response is.
- The buttons really are bulky. Let's swap their order and replace them with icons. The "Fix" button should have an edit icon (the pencil), and the "Confirm" button should have a tick mark and be colored green. The buttons should have the tooltips on-hover: "Edit this response" and "Confirm this response".
- Can we have some kind of temporarily-appearing green "toast" with a check mark when an answer is confirmed, similar to when an exploration is saved?
- When the creator confirms a fix, a similar toast should also appear.
- The exit button on the "Fix" modal should not be green.
- The exit buttons on both modals should not have grey background, either. Look at e.g. the Cancel button on the state content editor for the appropriate styling.
- In the "Fix" modal, there should be different indications for "selected" and "on-hover". They look the same at the moment. Perhaps the currently selected one could have a different background, and/or a highlighted border around it. The on-hover action could be more subtle.
- In the "Fix" modal, the feedback option that Oppia naturally gives at the moment should start off being highlighted as "currently selected".
- In the "Fix" modal, the Confirm button should be placed to the right of the Exit button on the bottom. The layout should be similar to the state content editor.
- On reflection, let's have the fix modal be fixed height, and have scrolling of responses within the "select a correct feedback" section, with a fixed max-height for that section. This allows the answer to always be in view. Also, with the current implementation, on the second card of the Welcome exploration, I ended up clicking the first piece of feedback repeatedly before realizing I had to go all the way to the bottom to confirm it. Some users may never even see the bottom.
- When I go to the second card of the "Welcome" exploration, open an answer group, click "Modify Training Data", and click the "Add New Answer" tab, the state in the background switches to the first state unexpectedly. This seems to always happen when I do this sequence of actions.
- In the Training Data window, I'm not sure why we have two tabs -- it looks a bit ugly. You could have just one pane, with an "Add New Answer" form in a box at the top, and the training data below that. The training data could be in a fixed-max-height div, similar to what was suggested above for the responses in the Fix modal.
- In the Training Data window, the Exit button should not be green (see comments above).
- Entering a sample answer in the "add new answer" tab does nothing, and a "TypeError: interactionRulesService[rule.type] is not a function" error is displayed in the browser console. (We might want to add an issue to add e2e tests for this...)

prasanna08 committed:
Addressed UI review comments

prasanna08 commented:
Hi @seanlip,

I have addressed all the comments you have in UI except following ones.

> When I go to the second card of the "Welcome" exploration, open an answer group, click "Modify Training Data", and click the "Add New Answer" tab, the state in the background switches to the first state unexpectedly. This seems to always happen when I do this sequence of actions.

Unable to reproduce this one. I tried loading welcome from admin and did what you did but nothing abnormal happened. Also I've modified UI of training data editor from 2 tabs to single continuous thing as you suggested so there is no new answer tab anymore. Maybe that solved error, IDK. Check it out.

> Entering a sample answer in the "add new answer" tab does nothing, and a "TypeError: interactionRulesService[rule.type] is not a function" error is displayed in the browser console. (We might want to add an issue to add e2e tests for this...)

I tried adding some sample answers in 2nd state of welcome exploration and it worked fine. It added the new answers to training data which you can see at the bottom in Training Data Answers section.

prasanna08 committed:
Merge branch 'develop' into training_interface

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/TrainingDataEditorPanelService.js:
```diff
@@ -146,7 +149,9 @@ oppia.factory('TrainingDataEditorPanelService', [
               $scope.openTrainUnresolvedAnswerModal = function(answerIndex) {
-                if ($scope.trainingData.length > 1) {
+                if (($scope.answerGroupHasNonEmptyRules &&
+                    $scope.trainingData.length > 0) ||
```
Not sure I understand this criterion. Could you explain (maybe in a comment)?

prasanna08 replied:
```
// An answer group must have either a rule or at least one
// answer in training data. Don't allow modification of training
// data answers if there are no rules and only one training data
// answer is present.
```

seanlip replied:
Ah ok, that's a good comment, thanks!

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/UnresolvedAnswersOverviewDirective.js:
```diff
@@ -150,6 +152,14 @@ oppia.directive('unresolvedAnswersOverview', [
                   var rulesService = $injector.get(rulesServiceName);
 
+                  var RESPONSE_CONFIRMED_TOAST_MESSAGE = (
+                    'Response of the answer has been confirmed');
```
`The answer "{{short snippet of answer}}..." has been successfully trained.`

(If the answer is sufficiently short, don't need the "...". But including a snippet helps distinguish the messages for different answers, especially if the creator is working quite quickly.

prasanna08 replied:
The answer snippet would work for interactions like text, numeric input only. It wouldn't work for code repl and similar complex input interaction.

prasanna08 replied:
Done

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/UnresolvedAnswersOverviewDirective.js:
```diff
@@ -150,6 +152,14 @@ oppia.directive('unresolvedAnswersOverview', [
+                  var RESPONSE_CONFIRMED_TOAST_MESSAGE = (
+                    'Response of the answer has been confirmed');
+                  var RESPONSE_FIXED_TOAST_MESSAGE = (
+                    'Response of the answer has been fixed');
```
`The response for "{{short snippet of answer}}..." has been fixed.`

prasanna08 replied:
Done

seanlip commented on feconf.py:
```diff
@@ -718,7 +718,7 @@ def get_empty_ratings():
 
 # Unfinished features.
-SHOW_TRAINABLE_UNRESOLVED_ANSWERS = False
+SHOW_TRAINABLE_UNRESOLVED_ANSWERS = True
```
Make sure to revert changes to feconf.

prasanna08 replied:
Done

seanlip commented on feconf.py:
```diff
@@ -892,7 +892,7 @@ def get_empty_ratings():
 QUESTION_BATCH_SIZE = 10
 
-STATE_ANSWER_STATS_MIN_FREQUENCY = 2
+STATE_ANSWER_STATS_MIN_FREQUENCY = 1
```
Ditto.

prasanna08 replied:
Done

seanlip reviewed (changes requested):
Make sure to revert changes to feconf.

UI comments coming soon, but I think you might also want to ask @pranavsid98 and @brianrodri to poke around at this as well and leave feedback!

seanlip commented:
More UI feedback:
- At the top of the "Confirm Oppia's responses" modal, let's give the user some context. In italics: "Oppia was unable to determine which response to give to the following student answers. Help Oppia by clicking the 'tick mark' button to confirm a response, or the 'edit' button to select a better response!"
- The button tooltips take too long to show up. They really should show up instantly, so let's use an actual tooltip (bootstrap) rather than a title attr. Also, the "Confirm this response" tooltip is misspelled.
- In the "fix" modal, only the list of Oppia's responses should be scrollable, not the entire modal contents. Otherwise I can't see the student's answer when selecting a response. Also (at least on Chrome on Mac), the scrollbar isn't obvious, I thought the responses cut off after the second one when in fact they were scrollable.
- In the "fix" modal, put the Exit button to the left of Confirm, not on the right. Follow the pattern used in the state content editor.
- "No feedback specified" seems to appear when I try to type a new response in the modal. That shouldn't happen.
- In the "fix" modal, change the text of the confirm button to "Confirm Response", and the text of the "Exit" button to "Cancel".
- Go to any answer group and add a new answer. Then move that answer to a different group. You still see, in the modal, "The answer is successfully added to training data." Have that latter message be a toast instead.

seanlip commented:
I think for code input it is still OK to show a small snippet of the start
of the code. We can handle other cases later.

On Tuesday, June 19, 2018, Prasanna Patil <notifications@github.com> wrote:

> *@prasanna08* commented on this pull request.
> ------------------------------
>
> In core/templates/dev/head/pages/exploration_editor/editor_tab/
> UnresolvedAnswersOverviewDirective.js
> <https://github.com/oppia/oppia/pull/5098#discussion_r196393621>:
>
> > @@ -150,6 +152,14 @@ oppia.directive('unresolvedAnswersOverview', [
>                    // Inject RulesService dynamically.
>                    var rulesService = $injector.get(rulesServiceName);
>
> +                  var RESPONSE_CONFIRMED_TOAST_MESSAGE = (
> +                    'Response of the answer has been confirmed');
>
> The answer snippet would work for interactions like text, numeric input
> only. It wouldn't work for code repl and similar complex input interaction.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/5098#discussion_r196393621>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyuQbUo-0_eNK31vuSWOWo2hwGBhNks5t-ObYgaJpZM4Ula_x>
> .
>

prasanna08 commented:
UI comments:
> At the top of the "Confirm Oppia's responses" modal, let's give the user some context. In italics: "Oppia was unable to determine which response to give to the following student answers. Help Oppia by clicking the 'tick mark' button to confirm a response, or the 'edit' button to select a better response!"

Done.

>The button tooltips take too long to show up. They really should show up instantly, so let's use an actual tooltip (bootstrap) rather than a title attr. Also, the "Confirm this response" tooltip is misspelled.

Done.

> In the "fix" modal, only the list of Oppia's responses should be scrollable, not the entire modal contents. Otherwise I can't see the student's answer when selecting a response. Also (at least on Chrome on Mac), the scrollbar isn't obvious, I thought the responses cut off after the second one when in fact they were scrollable.

Done. Also, the scrollbar is permanently visible now.

> In the "fix" modal, put the Exit button to the left of Confirm, not on the right. Follow the pattern used in the state content editor.

Done.

> "No feedback specified" seems to appear when I try to type a new response in the modal. That shouldn't happen.

This is not possible in current implementation as we are using `outcome-feedback-editor` directive to get a new response and that string is part of that directive. So, if I modify outcome-feedback-editor directive then all other places where it is used will also get modified (like answer group editor). So should I modify the directive?

> In the "fix" modal, change the text of the confirm button to "Confirm Response", and the text of the "Exit" button to "Cancel".

Done.

> Go to any answer group and add a new answer. Then move that answer to a different group. You still see, in the modal, "The answer is successfully added to training data." Have that latter message be a toast instead.

Done.

prasanna08 committed:
Addressed review comments

seanlip reviewed (approved):
HI @prasanna08 -- sorry, I can't look at this in detail at the moment, and I don't want to block you! Feel free to submit if you're happy with the UI (and ideally if you've got some confirmatory feedback on this from Pranav/Brian as well).

Thanks!

prasanna08 committed:
Merge branch 'develop' of github.com:oppia/oppia into training_interface

prasanna08 commented:
@seanlip, @brianrodri, @pranavsid98 I am going to merge this PR once all tests pass. Do any of you have any further review comments?

prasanna08 committed:
Fix backend tests

seanlip commented:
Hi @prasanna08, sorry for late reply! I am going to give this one more poke tonight (probably in the next couple of hours) -- stay tuned.

seanlip commented:
Hi @prasanna08 -- FYI, there are still merge conflicts?

prasanna08 commented:
Seems like this one surfaced very recently. I merged with develop just yesterday.

prasanna08 committed:
Merge branch 'develop' into training_interface

seanlip commented:
More UI comments!

- In the training data modal that appears when editing training data:
  - If there is nothing entered, the Submit button should be disabled (and not green).
  - Once a new answer has been successfully submitted, the "add new answer" field should be cleared.
  - When hovering over a training data answer, the "answer tile" should highlight to make clearer that it's selectable.
  - In the "Select Oppia's Response" dialog, when the "New Response" button is clicked, the "Confirm Response" button at the bottom of the modal should be disabled until the new response field is saved or cancelled. (There shouldn't be more than one active green button at a time.)
  - The "New Response" button should be renamed to "Add New Response" or "Create New Response".

I also have a further suggestion on the training experience in general, but that is a larger change so I am fine leaving it for a separate PR. I'll file an issue for this.

Thanks!

prasanna08 committed:
Addressed review comments

prasanna08 commented:
Hi, @seanlip 
All comments have been addressed except
> Once a new answer has been successfully submitted, the "add new answer" field should be cleared.

It is not possible to directly clear the answer because of none of the interaction set `$scope.answer` to its default value after submission. So it has to be modified for each interaction separately (and no single line fix is possible). Although fix itself is easy enough. All that has to be done is to set `$scope.answer=initiaDefaultlValue`.

seanlip commented:
Thanks @prasanna08! OK, then, what I suggest is, let's go ahead and get this in -- it's been sitting out for a while. 

I think the answer thing is still confusing, but would it be OK to file an issue for it in a separate PR? (I assume that resetting the answer won't have any negative effect on other places where the interaction is used, right?) Then, once this is merged, there will be a baseline to work from for both this issue and #5183, which means that some parallelization can happen.

Thanks!

seanlip commented:
(Also, feel free to merge this into develop if you're happy with it.)

prasanna08 commented:
> I think the answer thing is still confusing, but would it be OK to file an issue for it in a separate PR? (I assume that resetting the answer won't have any negative effect on other places where the interaction is used, right?) Then, once this is merged, there will be a baseline to work from for both this issue and #5183, which means that some parallelization can happen.

No, setting answer to default value wouldn't affect any other places. In fact, that should have been done from beginning. The only reason I can think of it has not been done is that because after the student enters answer the conversation moves to response and so resetting the input field to default value wasn't required. So it would not have any apparent effects on the conversation. I'll file an issue.

hoangviet1993 commented on core/templates/dev/head/components/answer_group_editor_directive.html:
```diff
@@ -82,6 +86,12 @@
+
+    <div ng-if="isCurrentInteractionTrainable()">
+      <button type="button" class="btn btn-default btn-lg oppia-add-rule-button protractor-test-add-answer oppia-modify-training-data-button" ng-click="openTrainingDataEditor()">
+        <i class="material-icons md-24">playlist_add</i> Modify Training Data
```
@prasanna08 I am getting some warnings from my e2e of duplicated "protractor-test-add-answer" button. Is it intended?

apb7 replied:
Hi @hoangviet1993, are you getting something like: Duplicate <class selector> found. Using the first occurence.
If this is the case, then there is no problem IMO.

hoangviet1993 replied:
I think the tag should be `protractor-test-modify-training-data-button`. The test was expecting "add-answer" button to not be clickable but the 'Modify training' button is clickable due to incorrect tag.

seanlip replied:
No idea about the tag (that's a question for @prasanna08) but just to note that we shouldn't get the sort of message that @apb7 is describing -- if there are multiple elements with the same class then we should be expecting a list and explicitly using .first(), otherwise we should ensure that the element with that class is unique.

prasanna08 replied:
@hoangviet1993, no that wasn't intended. I think it is happening bacause I am using same class as the "Add new answer group" button above. Can you file an issue for this? I will fix it asap. Just have to create new class for "Modify training data" button which has same style aa current class.

hoangviet1993 replied:
@prasanna08 #5204 is created per your request, PTAL 👍
